File: spark/src/main/scala/core/wc/Spark01_WordCount.scala
package core.wc

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

object Spark01_WordCount {
  def main(args: Array[String]): Unit = {
    //MC TODO:建立和Spark框架的连接
    val sparkConf = new SparkConf().setMaster("local").setAppName("WordCount")
    val context = new SparkContext(sparkConf)
    //MC TODO:执行业务操作
    //MC TODO:读取文件
    var lines: RDD[String] = context.textFile("F:\\大数据资料\\learning\\BigData-learning\\data\\data.txt")
    //MC TODO:将数据进行拆分
    val words = lines.flatMap(_.split(" "))
    //MC TODO:将数据根据单词进行分组 便于统计
    val wordGroup = words.groupBy(word => word)
    //MC TODO:分组后数据进行转换
    //    (hello, hello, hello), (world, world)
    //    (hello, 3), (world, 2)
    val wordToCount = wordGroup.map {
      case (word,list) => {
        (word,list.size)
      }
    }
    //MC TODO:打印
    val arrays: Array[(String, Int)] = wordToCount.collect()
    arrays.foreach(println)
    //MC TODO:关闭连接
    context.stop()
  }
}

File: spark/src/main/java/com/meicilly/core/wc/Spark01_WordCount.java
package com.meicilly.core.wc;

import org.apache.spark.SparkConf;
import org.apache.spark.SparkContext;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.*;
import org.apache.spark.rdd.RDD;
import scala.Tuple2;


import java.util.Arrays;
import java.util.Iterator;

public class Spark01_WordCount {
    public static void main(String[] args) {
        //MC TODO:构建连接
        SparkConf sparkConf = new SparkConf().setMaster("local").setAppName("WordCount");
        JavaSparkContext sc = new JavaSparkContext(sparkConf);
        //MC TODO:执行业务操作
        //MC TODO: 1.读取文件
        JavaRDD<String> linesRDD = sc.textFile("F:\\大数据资料\\learning\\BigData-learning\\data\\data.txt");

        //MC TODO: 2.将一行数据进行拆分 形成一个一个的单词
        JavaRDD<String> wordFlatmap = linesRDD.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public Iterator<String> call(String line) throws Exception {
                return Arrays.asList(line.split(" ")).iterator();
            }
        });
        JavaPairRDD<String, Integer> javaPairRDD = wordFlatmap.mapToPair(new PairFunction<String, String, Integer>() {
            @Override
            public Tuple2<String, Integer> call(String s) throws Exception {
                return new Tuple2<String, Integer>(s, 1);
            }
        });
        JavaPairRDD<String, Integer> result = javaPairRDD.reduceByKey(new Function2<Integer, Integer, Integer>() {
            @Override
            public Integer call(Integer v1, Integer v2) throws Exception {
                return v1 + v2;
            }
        });
            result.foreach(new VoidFunction<Tuple2<String, Integer>>() {
            @Override
            public void call(Tuple2<String, Integer> tuple) throws Exception {
                //System.out.println(tuple);
                System.out.println(tuple._1 + " " + tuple._2);
            }
        });
        sc.close();

    }
}
